File: src/agent_sdk/types_test.py
from __future__ import annotations

import json
from uuid import uuid4

import pytest
from pydantic import BaseModel

from agent_sdk.types import (
    AgentError,
    AudioResult,
    Capability,
    ErrorKind,
    ImageResult,
    Message,
    ModelInfo,
    Response,
    StructuredResponse,
    Tier,
    parse_json_from_llm,
)
from pathlib import Path


# ##################################################################
# tier values
# verify all five tiers have the expected string values
def test_tier_values() -> None:
    assert Tier.HIGH.value == "high"
    assert Tier.MEDIUM.value == "medium"
    assert Tier.LOW.value == "low"
    assert Tier.FREE_FAST.value == "free_fast"
    assert Tier.FREE_THINKING.value == "free_thinking"


# ##################################################################
# capability values
# verify all six capabilities exist
def test_capability_values() -> None:
    assert len(Capability) == 6
    assert Capability.TEXT.value == "text"
    assert Capability.IMAGE.value == "image"


# ##################################################################
# error kind values
# verify all error classifications exist
def test_error_kind_values() -> None:
    assert len(ErrorKind) == 6
    assert ErrorKind.RATE_LIMIT.value == "rate_limit"
    assert ErrorKind.AUTH.value == "auth"


# ##################################################################
# model info creation and qualified name
# verify frozen dataclass and property
def test_model_info_qualified_name() -> None:
    info = ModelInfo(
        provider="claude",
        model_id="claude-opus-4-6",
        display_name="Claude Opus",
        capabilities=frozenset({Capability.TEXT, Capability.STRUCTURED}),
        tier=Tier.HIGH,
    )
    assert info.qualified_name == "claude:claude-opus-4-6"
    assert info.supports_streaming is True
    assert info.supports_tools is False


# ##################################################################
# model info is frozen
# confirm immutability
def test_model_info_frozen() -> None:
    info = ModelInfo(
        provider="ollama",
        model_id="qwen3-8b",
        display_name="Qwen3 8B",
        capabilities=frozenset({Capability.TEXT}),
        tier=Tier.FREE_FAST,
    )
    with pytest.raises(AttributeError):
        info.provider = "changed"  # type: ignore[misc]


# ##################################################################
# message to dict
# verify serialisation with and without metadata
def test_message_to_dict_without_metadata() -> None:
    msg = Message(role="user", content="hello")
    d = msg.to_dict()
    assert d == {"role": "user", "content": "hello"}
    assert "metadata" not in d


def test_message_to_dict_with_metadata() -> None:
    msg = Message(role="assistant", content="hi", metadata={"tokens": 5})
    d = msg.to_dict()
    assert d["metadata"] == {"tokens": 5}


# ##################################################################
# response fields
# verify all fields are accessible
def test_response_fields() -> None:
    model = ModelInfo(
        provider="test",
        model_id="test-1",
        display_name="Test",
        capabilities=frozenset({Capability.TEXT}),
        tier=Tier.LOW,
    )
    cid = uuid4()
    tid = uuid4()
    resp = Response(text="answer", model_used=model, conversation_id=cid, turn_id=tid)
    assert resp.text == "answer"
    assert resp.model_used.provider == "test"
    assert resp.usage == {}


# ##################################################################
# structured response
# verify parsed field carries pydantic instance
def test_structured_response() -> None:
    class Score(BaseModel):
        value: int

    model = ModelInfo(
        provider="test",
        model_id="test-1",
        display_name="Test",
        capabilities=frozenset({Capability.STRUCTURED}),
        tier=Tier.LOW,
    )
    parsed = Score(value=42)
    resp = StructuredResponse(
        text='{"value": 42}',
        model_used=model,
        conversation_id=uuid4(),
        turn_id=uuid4(),
        parsed=parsed,
    )
    assert resp.parsed.value == 42
    assert isinstance(resp, Response)


# ##################################################################
# image result fields
# verify path and dimensions
def test_image_result() -> None:
    model = ModelInfo(
        provider="local",
        model_id="z-image-turbo",
        display_name="Z-Image Turbo",
        capabilities=frozenset({Capability.IMAGE}),
        tier=Tier.HIGH,
    )
    result = ImageResult(
        path=Path("/tmp/test.jpg"),
        model_used=model,
        conversation_id=uuid4(),
        prompt="a cat",
        width=512,
        height=512,
    )
    assert result.width == 512
    assert result.prompt == "a cat"


# ##################################################################
# audio result fields
# verify voice and optional duration
def test_audio_result() -> None:
    model = ModelInfo(
        provider="local",
        model_id="qwen3-tts",
        display_name="Qwen3 TTS",
        capabilities=frozenset({Capability.TTS}),
        tier=Tier.HIGH,
    )
    result = AudioResult(
        path=Path("/tmp/test.mp3"),
        model_used=model,
        conversation_id=uuid4(),
        text="hello world",
        voice="gary",
    )
    assert result.duration_seconds is None
    assert result.voice == "gary"


# ##################################################################
# agent error
# verify structured error with attempts
def test_agent_error_to_dict() -> None:
    err = AgentError(
        "rate limited on all providers",
        kind=ErrorKind.RATE_LIMIT,
        attempts=[
            {"provider": "claude", "status": 429},
            {"provider": "codex", "status": 429},
        ],
    )
    d = err.to_dict()
    assert d["kind"] == "rate_limit"
    assert len(d["attempts"]) == 2
    assert "rate limited" in d["error"]


def test_agent_error_is_exception() -> None:
    err = AgentError("test", kind=ErrorKind.INTERNAL)
    with pytest.raises(AgentError):
        raise err


# ##################################################################
# parse json from llm
# verify markdown code block stripping and plain json
def test_parse_json_plain() -> None:
    result = parse_json_from_llm('{"key": "value"}')
    assert result == {"key": "value"}


def test_parse_json_with_code_block() -> None:
    text = '```json\n{"key": "value"}\n```'
    result = parse_json_from_llm(text)
    assert result == {"key": "value"}


def test_parse_json_with_bare_code_block() -> None:
    text = '```\n{"items": [1, 2, 3]}\n```'
    result = parse_json_from_llm(text)
    assert result == {"items": [1, 2, 3]}


def test_parse_json_invalid_raises() -> None:
    with pytest.raises(json.JSONDecodeError):
        parse_json_from_llm("not json at all")


def test_parse_json_with_whitespace() -> None:
    text = '  \n  {"a": 1}  \n  '
    result = parse_json_from_llm(text)
    assert result == {"a": 1}

File: go/types_test.go
package dazagentsdk

import (
	"testing"
)

func TestTierValues(t *testing.T) {
	tests := []struct {
		tier Tier
		want string
	}{
		{TierVeryHigh, "very_high"},
		{TierHigh, "high"},
		{TierMedium, "medium"},
		{TierLow, "low"},
		{TierFreeFast, "free_fast"},
		{TierFreeThinking, "free_thinking"},
	}
	for _, tt := range tests {
		if string(tt.tier) != tt.want {
			t.Errorf("Tier %v = %q, want %q", tt.tier, string(tt.tier), tt.want)
		}
	}
}

func TestCapabilityValues(t *testing.T) {
	tests := []struct {
		cap  Capability
		want string
	}{
		{CapabilityText, "text"},
		{CapabilityStructured, "structured"},
		{CapabilityAgentic, "agentic"},
		{CapabilityImage, "image"},
		{CapabilityTTS, "tts"},
		{CapabilitySTT, "stt"},
	}
	for _, tt := range tests {
		if string(tt.cap) != tt.want {
			t.Errorf("Capability %v = %q, want %q", tt.cap, string(tt.cap), tt.want)
		}
	}
}

func TestErrorKindValues(t *testing.T) {
	tests := []struct {
		kind ErrorKind
		want string
	}{
		{ErrorRateLimit, "rate_limit"},
		{ErrorAuth, "auth"},
		{ErrorTimeout, "timeout"},
		{ErrorInvalidRequest, "invalid_request"},
		{ErrorNotAvailable, "not_available"},
		{ErrorInternal, "internal"},
	}
	for _, tt := range tests {
		if string(tt.kind) != tt.want {
			t.Errorf("ErrorKind %v = %q, want %q", tt.kind, string(tt.kind), tt.want)
		}
	}
}

func TestModelInfoQualifiedName(t *testing.T) {
	m := ModelInfo{
		Provider:    "claude",
		ModelID:     "claude-opus-4-6",
		DisplayName: "Claude Opus",
		Tier:        TierHigh,
	}
	want := "claude:claude-opus-4-6"
	if got := m.QualifiedName(); got != want {
		t.Errorf("QualifiedName() = %q, want %q", got, want)
	}
}

func TestMessageToDict(t *testing.T) {
	t.Run("without metadata", func(t *testing.T) {
		m := Message{Role: "user", Content: "hello"}
		d := m.ToDict()
		if d["role"] != "user" {
			t.Errorf("role = %v, want user", d["role"])
		}
		if d["content"] != "hello" {
			t.Errorf("content = %v, want hello", d["content"])
		}
		if _, ok := d["metadata"]; ok {
			t.Error("metadata should not be present when empty")
		}
	})

	t.Run("with metadata", func(t *testing.T) {
		m := Message{
			Role:     "assistant",
			Content:  "hi",
			Metadata: map[string]any{"key": "val"},
		}
		d := m.ToDict()
		if d["role"] != "assistant" {
			t.Errorf("role = %v, want assistant", d["role"])
		}
		meta, ok := d["metadata"].(map[string]any)
		if !ok {
			t.Fatal("metadata should be present")
		}
		if meta["key"] != "val" {
			t.Errorf("metadata[key] = %v, want val", meta["key"])
		}
	})
}

func TestAgentError(t *testing.T) {
	err := NewAgentError("something broke", ErrorInternal, nil)
	if err.Error() != "something broke" {
		t.Errorf("Error() = %q, want %q", err.Error(), "something broke")
	}
	if err.Kind != ErrorInternal {
		t.Errorf("Kind = %v, want %v", err.Kind, ErrorInternal)
	}
	if len(err.Attempts) != 0 {
		t.Errorf("Attempts should be empty, got %d", len(err.Attempts))
	}

	d := err.ToDict()
	if d["error"] != "something broke" {
		t.Errorf("ToDict error = %v", d["error"])
	}
	if d["kind"] != "internal" {
		t.Errorf("ToDict kind = %v", d["kind"])
	}
}

func TestAgentErrorWithAttempts(t *testing.T) {
	attempts := []map[string]any{
		{"provider": "claude", "error": "rate limited"},
	}
	err := NewAgentError("all failed", ErrorRateLimit, attempts)
	if len(err.Attempts) != 1 {
		t.Fatalf("Attempts = %d, want 1", len(err.Attempts))
	}
	if err.Attempts[0]["provider"] != "claude" {
		t.Errorf("attempt provider = %v", err.Attempts[0]["provider"])
	}
}

func TestParseJSONFromLLM_Plain(t *testing.T) {
	input := `{"name": "test", "value": 42}`
	result, err := ParseJSONFromLLM(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["name"] != "test" {
		t.Errorf("name = %v, want test", result["name"])
	}
	if result["value"] != float64(42) {
		t.Errorf("value = %v, want 42", result["value"])
	}
}

func TestParseJSONFromLLM_WithFences(t *testing.T) {
	input := "```json\n{\"key\": \"val\"}\n```"
	result, err := ParseJSONFromLLM(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["key"] != "val" {
		t.Errorf("key = %v, want val", result["key"])
	}
}

func TestParseJSONFromLLM_WithFencesNoLang(t *testing.T) {
	input := "```\n{\"key\": \"val\"}\n```"
	result, err := ParseJSONFromLLM(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["key"] != "val" {
		t.Errorf("key = %v, want val", result["key"])
	}
}

func TestParseJSONFromLLM_WithProse(t *testing.T) {
	input := "Here is the result of my analysis:\n\n{\"answer\": \"yes\"}"
	result, err := ParseJSONFromLLM(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["answer"] != "yes" {
		t.Errorf("answer = %v, want yes", result["answer"])
	}
}

func TestParseJSONFromLLM_ProseWithFencedBlock(t *testing.T) {
	input := "I analyzed the data and here are my findings:\n\n```json\n{\"status\": \"ok\"}\n```\n\nThat's all."
	result, err := ParseJSONFromLLM(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["status"] != "ok" {
		t.Errorf("status = %v, want ok", result["status"])
	}
}

func TestParseJSONFromLLM_NoJSON(t *testing.T) {
	input := "This is just plain text with no JSON at all."
	_, err := ParseJSONFromLLM(input)
	if err == nil {
		t.Fatal("expected error for text with no JSON")
	}
}

func TestParseJSONFromLLM_Whitespace(t *testing.T) {
	input := "  \n  {\"trimmed\": true}  \n  "
	result, err := ParseJSONFromLLM(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["trimmed"] != true {
		t.Errorf("trimmed = %v, want true", result["trimmed"])
	}
}
